File: src/commands/astCommand.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast

(***********************************************************************)
(* flow ast command *)
(***********************************************************************)

type ast_file_type =
  | File_json
  | File_js

type ast_include_comments =
  | Comments_true
  | Comments_false
  | Comments_docblock

let spec =
  {
    CommandSpec.name = "ast";
    doc = "Print the AST";
    usage =
      Printf.sprintf
        "Usage: %s ast [OPTION]... [FILE]\n\ne.g. %s ast foo.js\nor   %s ast < foo.js\n"
        CommandUtils.exe_name
        CommandUtils.exe_name
        CommandUtils.exe_name;
    args =
      CommandSpec.ArgSpec.(
        empty
        |> flag "--tokens" no_arg ~doc:"Include a list of syntax tokens in the output"
        |> flag "--pretty" no_arg ~doc:"Pretty-print JSON output"
        |> flag
             "--check"
             no_arg
             ~doc:"Checks whether the file parses, returning any errors but not the AST"
        |> flag "--debug" no_arg ~doc:"" (* undocumented *)
        |> flag
             "--pattern"
             no_arg
             ~doc:"Prints the AST structurally without locations to be used in pattern matching"
        |> flag
             "--type"
             (enum [("js", File_js); ("json", File_json)])
             ~doc:"Type of input file (js or json)"
        |> flag "--strict" no_arg ~doc:"Parse in strict mode"
        |> flag
             "--include-comments"
             (required
                ~default:Comments_true
                (enum
                   [
                     ("true", Comments_true);
                     ("false", Comments_false);
                     ("docblock-only", Comments_docblock);
                   ]
                )
             )
             ~doc:
               "Include or drop comments in the output (true, false or docblock-only) (default: true)"
        |> flag
             "--include-locs"
             (required ~default:true bool)
             ~doc:"Include or drop location information in the output (default: true)"
        |> CommandUtils.offset_style_flag
        |> CommandUtils.from_flag
        |> CommandUtils.path_flag
        |> anon "file" (optional string)
      );
  }

type ast_result_type =
  | Ast_json of (Loc.t, Loc.t) Ast.Expression.t
  | Ast_js of (Loc.t, Loc.t) Ast.Program.t

let get_file path = function
  | Some filename -> File_input.FileName (CommandUtils.expand_path filename)
  | None -> File_input.FileContent (path, Sys_utils.read_stdin_to_string ())

module Token_translator = Token_translator.Translate (Json_of_estree)

let pp_underscore_loc fmt _ = Format.pp_print_string fmt "_"

let main
    include_tokens
    pretty
    check
    debug
    pattern
    file_type_opt
    use_strict
    include_comments
    include_locs
    offset_style
    path
    filename
    () =
  let use_relative_path = Base.Option.value_map filename ~default:false ~f:Filename.is_relative in
  let file = get_file path filename in
  let content = File_input.content_of_file_input_unsafe file in
  let file_type =
    match file_type_opt with
    | Some t -> t
    | None ->
      begin
        match filename with
        | Some fn ->
          if Files.is_json_file fn then
            File_json
          else
            File_js
        | None -> File_js
      end
  in

  let module Translate =
    Estree_translator.Translate
      (Json_of_estree)
      (struct
        let include_locs = include_locs
      end)
  in
  (*
   * Record token stream into a list when the --tokens flag is passed.
   * Note that tokens stream in in order, so the list is constructed in reverse
   * order.
   *)
  let tokens = ref [] in
  let offset_kind = CommandUtils.offset_kind_of_offset_style offset_style in
  let offset_table = lazy (Offset_utils.make ~kind:offset_kind content) in
  let token_sink =
    if not include_tokens then
      None
    else
      Some
        (fun token_data ->
          tokens := Token_translator.token (Lazy.force offset_table) token_data :: !tokens)
  in
  Hh_json.(
    let results =
      try
        (* Make the parser as permissive as possible.
           TODO: make these CLI flags *)
        let parse_options =
          Some
            Parser_env.
              {
                enums = true;
                esproposal_class_instance_fields = true;
                esproposal_class_static_fields = true;
                esproposal_decorators = true;
                esproposal_export_star_as = true;
                esproposal_optional_chaining = true;
                esproposal_nullish_coalescing = true;
                types = true;
                use_strict;
              }
            
        in

        let filename = File_input.path_of_file_input file in
        let filename =
          if use_relative_path then
            Base.Option.map filename ~f:(Files.relative_path (Sys.getcwd ()))
          else
            filename
        in
        let (ast, errors) =
          match file_type with
          | File_js ->
            let filekey = Base.Option.map filename ~f:(fun s -> File_key.SourceFile s) in
            let (ocaml_ast, errors) =
              Parser_flow.program_file ~fail:false ~parse_options ~token_sink content filekey
            in
            if debug then (
              Ast.Program.pp Loc.pp Loc.pp Format.err_formatter ocaml_ast;
              Printf.eprintf "\n%!"
            );
            if pattern then (
              Ast.Program.pp pp_underscore_loc pp_underscore_loc Format.err_formatter ocaml_ast;
              Printf.eprintf "\n%!"
            );
            (Ast_js ocaml_ast, errors)
          | File_json ->
            let filekey = Base.Option.map filename ~f:(fun s -> File_key.JsonFile s) in
            let (ocaml_ast, errors) =
              Parser_flow.json_file ~fail:false ~parse_options ~token_sink content filekey
            in
            if debug then (
              Ast.Expression.pp Loc.pp Loc.pp Format.err_formatter ocaml_ast;
              Printf.eprintf "\n%!"
            );
            if pattern then (
              Ast.Expression.pp pp_underscore_loc pp_underscore_loc Format.err_formatter ocaml_ast;
              Printf.eprintf "\n%!"
            );
            (Ast_json ocaml_ast, errors)
        in
        if check then
          JSON_Object
            [("errors", Translate.errors errors); ("tokens", JSON_Array (List.rev !tokens))]
        else
          let offset_table = Some (Lazy.force offset_table) in
          let translated_ast =
            match ast with
            | Ast_js ast ->
              let ast =
                match include_comments with
                | Comments_true -> ast
                | Comments_false -> Comment_utils.strip_all_comments ast
                | Comments_docblock -> Comment_utils.strip_all_comments ~preserve_docblock:true ast
              in
              Translate.program offset_table ast
            | Ast_json ast ->
              let ast =
                match include_comments with
                | Comments_true -> ast
                | Comments_false
                | Comments_docblock ->
                  Comment_utils.strip_inlined_comments_expression ast
              in
              Translate.expression offset_table ast
          in
          match translated_ast with
          | JSON_Object params ->
            let errors_prop = ("errors", Translate.errors errors) in
            let tokens_prop = ("tokens", JSON_Array (List.rev !tokens)) in
            JSON_Object (errors_prop :: tokens_prop :: params)
          | _ -> assert false
      with
      | Parse_error.Error l -> JSON_Object [("errors", Translate.errors l)]
    in
    print_json_endline ~pretty results
  )

let command = CommandSpec.command spec main

File: rust_port/crates/flow_cli/src/ast_command.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::VecDeque;
use std::io::Read;
use std::path::Path;

use flow_parser::PERMISSIVE_PARSE_OPTIONS;
use flow_parser::ParseOptions;
use flow_parser::estree_translator;
use flow_parser::file_key::FileKey;
use flow_parser::file_key::FileKeyInner;
use flow_server_utils::file_input::FileInput;
use serde_json::json;

use crate::command_spec;
use crate::command_spec::arg_spec;
use crate::command_utils;

// ***********************************************************************
// flow ast command
// ***********************************************************************

#[derive(Clone)]
enum AstFileType {
    FileJson,
    FileJs,
}

#[derive(Clone)]
enum AstIncludeComments {
    CommentsTrue,
    CommentsFalse,
    CommentsDocblock,
}

fn spec() -> command_spec::Spec {
    let exe_name = command_utils::exe_name();
    let spec = command_spec::Spec::new(
        "ast",
        "Print the AST",
        format!(
            "Usage: {exe_name} ast [OPTION]... [FILE]\n\ne.g. {exe_name} ast foo.js\nor   {exe_name} ast < foo.js\n"
        ),
    )
    .flag(
        "--tokens",
        &arg_spec::truthy(),
        "Include a list of syntax tokens in the output",
        None,
    )
    .flag(
        "--pretty",
        &arg_spec::truthy(),
        "Pretty-print JSON output",
        None,
    )
    .flag(
        "--check",
        &arg_spec::truthy(),
        "Checks whether the file parses, returning any errors but not the AST",
        None,
    )
    .flag(
        "--debug",
        &arg_spec::truthy(),
        "", // undocumented
        None,
    )
    .flag(
        "--pattern",
        &arg_spec::truthy(),
        "Prints the AST structurally without locations to be used in pattern matching",
        None,
    )
    .flag(
        "--type",
        &arg_spec::optional(arg_spec::enum_flag(vec![
            ("js", AstFileType::FileJs),
            ("json", AstFileType::FileJson),
        ])),
        "Type of input file (js or json)",
        None,
    )
    .flag(
        "--strict",
        &arg_spec::truthy(),
        "Parse in strict mode",
        None,
    )
    .flag(
        "--no-enums",
        &arg_spec::truthy(),
        "Disable enum support",
        None,
    )
    .flag(
        "--no-component-syntax",
        &arg_spec::truthy(),
        "Disable support for component syntax",
        None,
    )
    .flag(
        "--include-comments",
        &arg_spec::required(
            Some(AstIncludeComments::CommentsTrue),
            arg_spec::enum_flag(vec![
                ("true", AstIncludeComments::CommentsTrue),
                ("false", AstIncludeComments::CommentsFalse),
                ("docblock-only", AstIncludeComments::CommentsDocblock),
            ]),
        ),
        "Include or drop comments in the output (true, false or docblock-only) (default: true)",
        None,
    )
    .flag(
        "--include-locs",
        &arg_spec::required(Some(true), arg_spec::bool_flag()),
        "Include or drop location information in the output (default: true)",
        None,
    );
    let spec = command_utils::add_offset_style_flag(spec);
    let spec = command_utils::add_from_flag(spec);
    let spec = command_utils::add_path_flag(spec);
    spec.anon("file", &arg_spec::optional(arg_spec::string()))
}

enum AstResultType {
    AstJson(flow_parser::ast::expression::Expression<flow_parser::loc::Loc, flow_parser::loc::Loc>),
    AstJs(flow_parser::ast::Program<flow_parser::loc::Loc, flow_parser::loc::Loc>),
}

fn get_file(path: Option<String>, filename: Option<String>) -> FileInput {
    match filename {
        Some(filename) => FileInput::FileName(command_utils::expand_path(&filename)),
        None => {
            let mut content = String::new();
            std::io::stdin()
                .read_to_string(&mut content)
                .expect("failed to read stdin");
            FileInput::FileContent(path, content)
        }
    }
}

fn offset(
    offset_table: &flow_parser::offset_utils::OffsetTable,
    position: flow_parser::loc::Position,
) -> usize {
    offset_table.offset(position).unwrap() as usize
}

fn translate_token(
    offset_table: &flow_parser::offset_utils::OffsetTable,
    token: &flow_parser::TokenSinkResult,
) -> serde_json::Value {
    let context = match token.token_context {
        flow_parser::LexMode::Normal => "normal",
        flow_parser::LexMode::Type => "type",
        flow_parser::LexMode::JsxTag => "jsxTag",
        flow_parser::LexMode::JsxChild => "jsxChild",
        flow_parser::LexMode::Regexp => "regexp",
    };
    json!({
        "type": token.token_kind.as_str(),
        "context": context,
        "loc": {
            "start": {
                "line": token.token_loc.start.line,
                "column": token.token_loc.start.column,
            },
            "end": {
                "line": token.token_loc.end.line,
                "column": token.token_loc.end.column,
            },
        },
        "range": [
            offset(offset_table, token.token_loc.start),
            offset(offset_table, token.token_loc.end),
        ],
        "value": token.token_kind.to_value(),
    })
}

fn main(args: &arg_spec::Values) {
    let include_tokens = command_spec::get(args, "--tokens", &arg_spec::truthy()).unwrap();
    let pretty = command_spec::get(args, "--pretty", &arg_spec::truthy()).unwrap();
    let check = command_spec::get(args, "--check", &arg_spec::truthy()).unwrap();
    let debug = command_spec::get(args, "--debug", &arg_spec::truthy()).unwrap();
    let pattern = command_spec::get(args, "--pattern", &arg_spec::truthy()).unwrap();
    let file_type_opt = command_spec::get(
        args,
        "--type",
        &arg_spec::optional(arg_spec::enum_flag(vec![
            ("js", AstFileType::FileJs),
            ("json", AstFileType::FileJson),
        ])),
    )
    .unwrap();
    let use_strict = command_spec::get(args, "--strict", &arg_spec::truthy()).unwrap();
    let no_enums = command_spec::get(args, "--no-enums", &arg_spec::truthy()).unwrap();
    let no_component_syntax =
        command_spec::get(args, "--no-component-syntax", &arg_spec::truthy()).unwrap();
    let include_comments = command_spec::get(
        args,
        "--include-comments",
        &arg_spec::required(
            Some(AstIncludeComments::CommentsTrue),
            arg_spec::enum_flag(vec![
                ("true", AstIncludeComments::CommentsTrue),
                ("false", AstIncludeComments::CommentsFalse),
                ("docblock-only", AstIncludeComments::CommentsDocblock),
            ]),
        ),
    )
    .unwrap();
    let include_locs = command_spec::get(
        args,
        "--include-locs",
        &arg_spec::required(Some(true), arg_spec::bool_flag()),
    )
    .unwrap();
    let offset_style = command_utils::get_offset_style(args);
    let path = command_spec::get(args, "--path", &arg_spec::optional(arg_spec::string())).unwrap();
    let filename =
        command_spec::get(args, "file", &arg_spec::optional(arg_spec::string())).unwrap();

    let use_relative_path = filename
        .as_ref()
        .is_some_and(|filename| Path::new(filename).is_relative());
    let file = get_file(path.clone(), filename.clone());
    let content = file.content_of_file_input_unsafe();
    let file_type = match file_type_opt {
        Some(t) => t,
        None => match filename.as_deref() {
            Some(filename) if flow_common::files::is_json_file(filename) => AstFileType::FileJson,
            Some(_) | None => AstFileType::FileJs,
        },
    };

    let offset_kind = command_utils::offset_kind_of_offset_style(offset_style);
    let offset_style = match offset_style {
        None | Some(command_utils::OffsetStyle::Utf8Bytes) => {
            estree_translator::OffsetStyle::Utf8Bytes
        }
        Some(command_utils::OffsetStyle::JsIndices) => estree_translator::OffsetStyle::JsIndices,
    };
    let config = estree_translator::Config {
        include_locs,
        include_filename: true,
        offset_style,
    };

    // Record token stream into a list when the --tokens flag is passed.
    // Note that tokens stream in in order, so the list is constructed in reverse
    // order.
    let mut tokens = VecDeque::new();
    let offset_table =
        flow_parser::offset_utils::OffsetTable::make_with_kind(offset_kind, &content);
    let mut token_sink = |token_data: flow_parser::TokenSinkResult| {
        tokens.push_front(translate_token(&offset_table, &token_data));
    };
    let token_sink = if !include_tokens {
        None
    } else {
        Some(&mut token_sink as &mut dyn FnMut(flow_parser::TokenSinkResult))
    };

    // Make the parser as permissive as possible.
    // TODO: make these CLI flags
    let parse_options = Some(ParseOptions {
        components: !no_component_syntax,
        enums: !no_enums,
        use_strict,
        ..PERMISSIVE_PARSE_OPTIONS
    });

    let filename = file.path_of_file_input().map(str::to_owned);
    let filename = if use_relative_path {
        filename.as_ref().map(|filename| {
            flow_common::files::relative_path(
                &std::env::current_dir().expect("failed to get current directory"),
                filename,
            )
        })
    } else {
        filename
    };

    let (ast, errors) = match file_type {
        AstFileType::FileJs => {
            // flow ast is a standalone parser - use the raw constructor
            // so paths are stored as-is without root stripping.
            let filekey = filename
                .clone()
                .map(|filename| FileKey::new(FileKeyInner::SourceFile(filename)));
            let (ocaml_ast, errors) = match filekey.clone() {
                Some(filekey) => flow_parser::parse_program_file::<()>(
                    false,
                    token_sink,
                    parse_options,
                    filekey,
                    Ok(&content),
                ),
                None => flow_parser::parse_program_without_file(
                    false,
                    token_sink,
                    parse_options,
                    Ok(&content),
                ),
            };

            if debug {
                eprintln!("{:?}", ocaml_ast);
            }
            if pattern {
                eprintln!("{:?}", ocaml_ast);
            }
            (AstResultType::AstJs(ocaml_ast), errors)
        }
        AstFileType::FileJson => {
            let filekey = filename
                .clone()
                .map(|filename| FileKey::new(FileKeyInner::JsonFile(filename)));
            let (ocaml_ast, errors) = flow_parser::parse_json_file(
                false,
                token_sink,
                parse_options,
                filekey.clone(),
                Ok(&content),
            );

            if debug {
                eprintln!("{:?}", ocaml_ast);
            }
            if pattern {
                eprintln!("{:?}", ocaml_ast);
            }
            (AstResultType::AstJson(ocaml_ast), errors)
        }
    };

    let tokens = tokens.into_iter().collect::<Vec<_>>();
    let results = if check {
        let mut obj = serde_json::Map::new();
        obj.insert(
            "errors".to_string(),
            estree_translator::errors(&offset_table, config.include_filename, &errors),
        );
        obj.insert("tokens".to_string(), serde_json::Value::Array(tokens));
        serde_json::Value::Object(obj)
    } else {
        let translated_ast = match ast {
            AstResultType::AstJs(mut ast) => {
                match include_comments {
                    AstIncludeComments::CommentsTrue => {}
                    AstIncludeComments::CommentsFalse => {
                        flow_parser::comment_utils::strip_all_comments(&mut ast, false);
                    }
                    AstIncludeComments::CommentsDocblock => {
                        flow_parser::comment_utils::strip_all_comments(&mut ast, true);
                    }
                }
                estree_translator::program(&offset_table, &config, &ast)
            }
            AstResultType::AstJson(mut ast) => {
                match include_comments {
                    AstIncludeComments::CommentsTrue => {}
                    AstIncludeComments::CommentsFalse | AstIncludeComments::CommentsDocblock => {
                        flow_parser::comment_utils::strip_inlined_comments_expression(&mut ast);
                    }
                }
                estree_translator::expression(&offset_table, &config, &ast)
            }
        };
        match translated_ast {
            serde_json::Value::Object(mut params) => {
                let errors_value =
                    estree_translator::errors(&offset_table, config.include_filename, &errors);
                let tokens_value = serde_json::Value::Array(tokens);
                let mut properties = serde_json::Map::new();
                properties.insert("errors".to_string(), errors_value);
                properties.insert("tokens".to_string(), tokens_value);
                properties.append(&mut params);
                serde_json::Value::Object(properties)
            }
            _ => panic!("expected translated_ast to be a JSON object"),
        }
    };
    flow_hh_json::print_json_endline(pretty, &results);
}

pub(crate) fn command() -> command_spec::Command {
    command_spec::command(spec(), main)
}
